cloudformation script for custom synapse bucket (#515)
Sage-Bionetworks/synapseDocs: articles/custom_storage_location.md

<br/>
For more information, please read: [How Do I Configure CORS on My Bucket?](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html#how-do-i-enable-cors){:target="_blank"}

### Setup with AWS Cloudformation

For convienance [AWS Cloudformation] can be used to provision a custom AWS S3 bucket for use with Synapse.
Using this approach will result in the exact same bucket as described in [Setup with AWS Console](#setup-with-aws-console).

Instructions:

> 1. Download the [CF template](https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/templates/SynapseExternalBucket-v2.yaml).

Author: We should make a copy of this script and add it to the docs instead providing  a reference to scicomp because the one in scicomp may disappear one day.  it's actually very, very likely to disappear.

Reviewer: Looks like it already disappeared. Do you still have a copy of the script?

Reviewer 2: Or we could document the process of how someone finds the right version of this file/script?

<br/>
For more information, please read: [How Do I Configure CORS on My Bucket?](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html#how-do-i-enable-cors){:target="_blank"}

> ### Setup bucket with AWS cloudformation

Reviewer 2: ```suggestion
### Setup bucket with AWS CloudFormation
```

> You can also use AWS Cloudformation to provision a custom Synapse bucket.

Reviewer 2: why would someone want to do this?

Who is paying for it - Sage, or end user? Who is in control of it?

Can the bucket be used for anything other than Synapse?

Author: Not sure if these are suggestions for more docs or your asking because you don't know but to answer:
> why would someone want to do this?
As noted in the commit message this is an alternative to creating the bucket manually.  if someone wants to run a single command to quickly and easily get a synapse bucket then they can use cloudformation which we conveniently already have a script for this.

> who is paying for it?
The end user, or the owner of the account that the bucket gets provisioned to.

> can the bucket be used for anything other than synapse
yes, it's just a generic AWS S3 bucket.  The only difference is that synapse has access to the bucket.

Reviewer 3: > As noted in the commit message this is an alternative to creating the bucket manually.

I think this should incorporate into the previous part of this documentation, explaining to users that there 2 ways to setup a bucket, and highlight the manual instructions with the same heading as the heading we use for using the CloudFormation script part.  

These answers are valuable to users as well. I had the same questions. So I think we should clearly address these.

With these understanding, I would immediately jump to the CloudFormation script and skip the manual instructions. Instead, with the current structure, I may follow all the manual instructions before reading the CF part. The heading does not make sense now, because this section has the same heading as a part of the manual instructions.

> You can also use AWS Cloudformation to provision a custom Synapse bucket.

Reviewer 2: ```suggestion
You can also use AWS CloudFormation to provision a custom AWS S3 bucket for use with Synapse.
```

Instructions:
1. Download the CF [template](https://github.com/Sage-Bionetworks/scicomp-provisioner/blob/master/templates/SynapseExternalBucket-v2.yaml).

> 2. Use the [awscli](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/index.html) or

Reviewer 2: ```suggestion
2. Use the [AWS Command Line](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/index.html) or
```

> [aws console](https://console.aws.amazon.com/cloudformation/home?region=us-east-1#/stacks/create) to execute the template which will provision the bucket.

Reviewer 2: ```suggestion
[AWS console](https://console.aws.amazon.com/cloudformation/home?region=us-east-1#/stacks/create) to execute the template which will provision the bucket.
```

> Example using the awscli:

Reviewer 2: ```suggestion
Example using `awscli`:
```

```
aws cloudformation create-stack \
--stack-name MyCustomSynapseBucket \
--template-body file://SynapseExternalBucket.yaml \

> --parameters ParameterKey=Department,ParameterValue=Cancer ParameterKey=Project,ParameterValue=Mammography \

Reviewer 2: why are these parameters mandatory? How do I determine them?

Author: No, reason just the way our template works.  They are just used to tag the bucket.  Tags are arbitrary.  Will add a link to cloudformation tags doc,  https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-resource-tags.html

Author: done

> ParameterKey=OwnerEmail,ParameterValue=joe.smith@company.com ParameterKey=AllowWriteBucket,ParameterValue=true

Reviewer 2: why is `ParameterKey=AllowWriteBucket,ParameterValue=true` required to fill out? wouldn't that be necessary in any case?

Author: This is to setup the bucket permission.  
AllowWriteBucket=true  means read-write bucket
AllowWriteBucket=false means read only bucket

ParameterKey=SynapseUserName,ParameterValue=jsmith
```

The above example shows required parameters.  The following are optional parameters:
```
# (Optional) true (default) to encrypt bucket, false for no encryption
EncryptBucket: 'false'
# (Optional) 'Enabled' to enable bucket versioning, default is 'Suspended'
BucketVersioning: 'Enabled'
# (Optional) 'Enabled' to enable bucket data life cycle rule, default is 'Disabled'
EnableDataLifeCycle: 'Enabled'
# (Optional) S3 bucket objects will transition into this storage class: GLACIER(default), STANDARD_IA, ONEZONE_IA
LifecycleDataStorageClass: 'STANDARD_IA'
# (Optional) Number of days until S3 objects are moved to the LifecycleDataStorageClass, default is 30
LifecycleDataTransition: '90'
# (Optional) Number of days (from creation) when objects are deleted from S3 and LifecycleDataStorageClass, default is 365000
LifecycleDataExpiration: '1825'
```

> 

Reviewer 2: How do I know it was successful? What do I do next?

Author: added info for this
